File: solutions/problem_2249_count_lattice_points_inside_a_circle/solution.py
"""
Problem 2249: Count Lattice Points Inside a Circle
================================================
Difficulty: Medium
Tags: Array, Hash Table, Math, Geometry, Enumeration
Pattern: Hash Map Lookup

Time Complexity:  O(n)
Space Complexity: O(n)
"""

class Solution:
    def countLatticePoints(self, circles: List[List[int]]) -> int:
        # Hash map approach - O(n) time, O(n) space
        seen = {}
        for i, val in enumerate(circles):
            complement = circles - val
            if complement in seen:
                return [seen[complement], i]
            seen[val] = i
        return 0

File: solutions/problem_2249_count_lattice_points_inside_a_circle/solution.cpp
/*
 * Problem 2249: Count Lattice Points Inside a Circle
 * ================================================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Math, Geometry, Enumeration
 * Pattern: Hash Map Lookup
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */

#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int countLatticePoints(vector<vector<int>>& circles) {
        // Hash map approach - O(n) time, O(n) space
        unordered_map<int, int> seen;
        for (int i = 0; i < circles.size(); i++) {
            int complement = circles - circles[i];
            if (seen.count(complement)) {
                return {seen[complement], i};
            }
            seen[circles[i]] = i;
        }
        return 0;
    }
};
